src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs

```diff
@@ -82,11 +71,81 @@ internal struct DiagnosticAnalysisResult
             IsEmpty = DocumentIds.IsEmpty && _others.IsEmpty;
         }
 
-        public DiagnosticAnalysisResult(DiagnosticAnalysisResultBuilder builder) :
-            this(builder.Project.Id, builder.Version,
-                builder.SyntaxLocals, builder.SemanticLocals, builder.NonLocals, builder.Others,
-                builder.DocumentIds, fromBuild: false)
+        public static DiagnosticAnalysisResult CreateEmpty(ProjectId projectId, VersionStamp version)
+        {
+            return new DiagnosticAnalysisResult(
+                projectId,
+                version,
+                documentIds: ImmutableHashSet<DocumentId>.Empty,
+                syntaxLocals: ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>>.Empty,
+                semanticLocals: ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>>.Empty,
+                nonLocals: ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>>.Empty,
+                others: ImmutableArray<DiagnosticData>.Empty,
+                fromBuild: false);
+        }
+
+        public static DiagnosticAnalysisResult CreateInitialResult(ProjectId projectId)
+        {
+            return new DiagnosticAnalysisResult(
+                projectId,
+                version: VersionStamp.Default,
+                documentIds: null,
+                isEmpty: true,
+                fromBuild: false);
+        }
+
+        public static DiagnosticAnalysisResult CreateFromBuild(Project project, ImmutableArray<DiagnosticData> diagnostics)
+        {
+            // we can't distinguish locals and non locals from build diagnostics nor determine right snapshot version for the build.
+            // so we put everything in as semantic local with default version. this lets us to replace those to live diagnostics when needed easily.
+            var version = VersionStamp.Default;
+
+            // filter out any document that doesn't support diagnostics
+            var group = diagnostics.GroupBy(d => d.DocumentId);
```

Comment:
so, we will put the diagnostic support check later?

Reply:
➡️ That is correct. The current pull request is just a refactoring. The `SupportsDiagnostics` method doesn't exist in the target branch yet, and this location will be minimally updated to use it when it's added.